Content.Shared/Doors/Systems/SharedDoorSystem.Airlock.cs

```diff
@@ -0,0 +1,189 @@
+            return;
+
+        args.Cancel();
+    }
+
+    private void OnBeforeDoorDenied(Entity<AirlockComponent> airlock, ref BeforeDoorDeniedEvent args)
+    {
+        if (CanChangeState(airlock))
+            return;
+
+        args.Cancel();
+    }
+
+    private void OnGetPryMod(EntityUid uid, AirlockComponent component, ref GetPryTimeModifierEvent args)
+    {
+        if (component.Powered)
+            args.PryTimeModifier *= component.PoweredPryModifier;
+
+        if (DoorSystem.IsBolted(uid))
+            args.PryTimeModifier *= component.BoltedPryModifier;
+    }
+
+    /// <summary>
+    /// Updates the auto close timer.
+    /// </summary>
+    public void UpdateAutoClose(Entity<AirlockComponent> airlock, DoorComponent? door = null)
+    {
+        if (!Resolve(airlock, ref door))
+            return;
+
+        if (door.State != DoorState.Open)
+            return;
+
+        if (!airlock.Comp.AutoClose)
+            return;
+
+        if (!CanChangeState(airlock))
+            return;
+
+        var autoCloseEvent = new BeforeDoorAutoCloseEvent();
+        RaiseLocalEvent(airlock, autoCloseEvent);
+
+        if (autoCloseEvent.Cancelled)
+            return;
+
+        DoorSystem.SetNextStateChange((airlock, door),
+            airlock.Comp.AutoCloseDelay * airlock.Comp.AutoCloseDelayModifier);
+    }
+
+    private void OnBeforePry(EntityUid uid, AirlockComponent component, ref BeforePryEvent args)
+    {
+        if (args.Cancelled)
+            return;
+
+        if (!component.Powered || args.PryPowered)
+            return;
+
+        args.Message = "airlock-component-cannot-pry-is-powered-message";
+
+        args.Cancelled = true;
+    }
+
+    public void UpdateEmergencyLightStatus(EntityUid uid, AirlockComponent component)
+    {
+        _appearance.SetData(uid, DoorVisuals.EmergencyLights, component.EmergencyAccess);
+    }
+
+    public void SetEmergencyAccess(Entity<AirlockComponent> airlock,
+        bool value,
+        EntityUid? user = null,
+        bool predicted = false)
+    {
+        if (!airlock.Comp.Powered)
+            return;
+
+        if (airlock.Comp.EmergencyAccess == value)
+            return;
+
+        airlock.Comp.EmergencyAccess = value;
+        Dirty(airlock, airlock.Comp); // This only runs on the server apparently, so we need this.
+        UpdateEmergencyLightStatus(airlock, airlock.Comp);
+
+        var sound = airlock.Comp.EmergencyAccess ? airlock.Comp.EmergencyOnSound : airlock.Comp.EmergencyOffSound;
+        if (predicted)
+            _audio.PlayPredicted(sound, airlock, user: user);
+        else
+            _audio.PlayPvs(sound, airlock);
+    }
+
+    public void SetAutoCloseDelayModifier(AirlockComponent component, float value)
+    {
+        if (component.AutoCloseDelayModifier.Equals(value))
+            return;
+
+        component.AutoCloseDelayModifier = value;
+    }
+
+    public void SetSafety(AirlockComponent component, bool value)
```

Comment:
```suggestion
    public static void SetSafety(AirlockComponent component, bool value)
```
May be worth considering making the component fields accessible to the door wire system based on this.